[트렌디 신재훈] Add: user 회원가입, 로그인, 토큰발행, product앱 SearchView 초안 작성

NoahShin opened the pull request:
유저 회원가입 view 작성
SignUpEmailView, SignUpIdView 역할:
아이디 작성후 탭을 누르거나 다음 입력칸으로 옮겼을 시에
validation가능, SignUpEmailView도 위와 동일

NoahShin committed:
Add: user 회원가입, 로그인, 토큰발행, product 검색view

유저 회원가입 view 작성
SignUpEmailView, SignUpIdView 역할:
아이디 작성후 탭을 누르거나 다음 입력칸으로 옮겼을 시에
validation가능, SignUpEmailView도 위와 동일

soheon-lee commented on core/utils.py:
```diff
@@ -0,0 +1,38 @@
+def login_decorator(func):
+    def wrapper(self, request, *args, **kwargs):
+        try:
+            token = request.headers.get('Authorization', None)
+            payload = jwt.decode(token, SECRET['secret'], algorithm=ALGORITHM[algorithm])
+            user = User.objects.get(id=payload['user_id'])
+            request.user = user
```
`=`을 기준으로 코드 한 번 정렬해주세요

NoahShin replied:
네!

soheon-lee commented on core/utils.py:
```diff
@@ -0,0 +1,38 @@
+    def wrapper(self, request, *args, **kwargs):
+        try:
+            token = request.headers.get('Authorization', None)
+            payload = jwt.decode(token, SECRET['secret'], algorithm=ALGORITHM[algorithm])
```
```suggestion
            payload = jwt.decode(token, SECRET['secret'], algorithm=ALGORITHM['algorithm'])
```
딕셔너리의 키는 아마도 string일 것입니다.
데코레이터를 아직 테스트 해보시기 전이라 티가 안났을텐데 아마 동작하지 않을거에요 확인부탁드립니다~

NoahShin replied:
확인했습니다. 감사합니다.

soheon-lee commented on product/models.py:
```diff
@@ -21,10 +21,13 @@ def __str__(self):
 
 class ProductDetail(core_models.TimeStampedModel):
-    image_url   = models.URLField(max_length=1000)
+    image_url_1 = models.URLField(max_length=1000,null=True)
+    image_url_2 = models.URLField(max_length=1000,null=True)
+    image_url_3 = models.URLField(max_length=1000,null=True)
+    image_url_4 = models.URLField(max_length=1000,null=True)
+    image_url_5 = models.URLField(max_length=1000,null=True)
```
product와 image를 1:N 관계로 선언하지 않고 모든 이미지의 url을 하나하나 컬럼으로 저장하시는 건가요?
만약 서비스에서 "앞으로 사진을 10장씩 저장합시다 :)" 라고 하면 확장성이 너무 떨어지게 됩니다. 1:N으로 바꾸시는 것을 추천드립니다. 지금 고생하시는 것이 나아요 👍

soheon-lee commented on product/views.py:
```diff
@@ -1,3 +1,25 @@
+
+from user.models import User 
+
+class SearchView(View):
+    def get(request):
+        data = json.loads(request.body) 
+        results = Product.objects.filter(title__contain=data['result'])
+
+        if not results.exists():
+            return JsonResponse({'MESSAGE':'NO_RESULT!'}, status = 400)
+    
+        results = [{
+            'title'            : product.title,
+            'thumb_image'      : product.thumb_image ,
+            'discounted_price' : round(product.price * (1-product.sale),-2),
+            'price'            : product.price,
+            'seller'           : product.seller.name,
+            'delivery'         : product.delivery,
+            'sale'             : str(product.sale * 100) + '%'
+            } for result in results]
+        return JsonResponse({'results': results}, status = 200)
```
이 엔드포인트 동작하나요? 확인해보셨나욤?

NoahShin replied:
수정하고, 검색 기능 동작 확인하고 다시 commit 했습니다.

soheon-lee commented on user/views.py:
```diff
@@ -1,3 +1,94 @@
+            return JsonResponse({'MESSAGE':'SUCCESS'}, status = 201)
+        except KeyError:
+            return JsonResponse({"MESSAGE" : "KEY_ERROR"}, status = 400)
+
+    def get(self, request):
+        user_data = User.objects.values()
+        return JsonResponse({'USER DATA' : list(user_data)}, status = 200)
```
가입된 사용자를 모두 불러오는 엔드포인트인가요?
이는 아무런 제약조건 없이 모든 사용자의 데이터를 전부 표출하기 때문에, 보안상의 이유로 실제로는 존재하면 안되는 기능입니다 ㅎㅎ
아깝지만 해당 함수를 제거해주세요~

NoahShin replied:
아 보안생각을 못했습니다. 네 제거 하겠습니다.

soheon-lee commented on user/views.py:
```diff
@@ -1,3 +1,94 @@
+            return JsonResponse({"MESSAGE" : "KEY_ERROR"}, status = 400)
+
+        if not nick_name:
+            return JsonResponse({"MESSAGE" : "NO_EXHIST_USER"}, status = 400)
```
오타 수정해주세요

NoahShin replied:
네 수정했습니다.

soheon-lee commented on user/views.py:
```diff
@@ -1,3 +1,94 @@
+        if not nick_name:
+            return JsonResponse({"MESSAGE" : "NO_EXHIST_USER"}, status = 400)
+
+        if User.objects.filter(nick_name=data['nick_name']).exists():
+            user = User.objects.get(nick_name=data['nick_name'])
+            if bcrypt.checkpw(data['password'].encode('UTF-8'), user.password.encode('UTF-8')):
+                token = jwt.encode({'user_id' : user.id}, SECRET['secret'], algorithm=ALGORITHM['algorithm']).decode('UTF-8')
+                return JsonResponse({'TOKEN' : token, 'user_name' : user.user_name}, status = 200)
+            return JsonResponse({"MESSAGE" : "INVALID_USER"}, status = 401)
```
goooood!

그런데 89번 if문이 만족하지 않을 때, 즉 주어진 nick_name을 가진 User가 존재하지 않을 때에는 어떤 response가 전달되나요?

NoahShin replied:
아.. 위쪽 코드에서  nick_name을 가진 user 가 존재하지 않을 시에 No exist 유저 response가 전달되도록해서 아래쪽에서 처리 할 생각을 못했습니다. 하나로 합치겠습니다. (if not 한번 멋있게 써보려다..ㅠ)

NoahShin committed:
Refactor 멘토 리뷰 반영 (1개 제외)
- product app안 views.py의 SearchView 구동되는지 확인 못했음
- 나머지 오타, 줄맞춤 수정
- User가 존재하지 않았을 시에 뿌릴 Response 로직 합침

NoahShin committed:
Modify: product앱의 검색기능View 구현
- 실험 안해보고 커밋했었는데, 이번엔 실험해보고 올림
- discounted price and sale 부분 datatype error 수정

soheon-lee commented on favor/views.py:
```diff
@@ -1,3 +1,65 @@
+            if ProductFavor.objects.filter(user=user_id, product=data['product_id']).exists():
+                ProductFavor.objects.get(user=user_id, product=data['product_id']).delete()
+                return JsonResponse({'MESSAGE':'REMOVED'}, status = 200)
+            
+            else:
+                ProductFavor(
+                    user_id    = user_id,
+                    product_id = data['product_id']
+                ).save()
+                return JsonResponse({'MESSAGE':'ADDED_TO_FAVOR_PRODUCT'}, status = 201)
```
else가 없어도 바로 위의 if문이 만족되지 않는다면 바로 도달할 수 있는 로직입니다. 꼭 필요한 경우가 아니면 else를 사용하지 않는 것이 보다 파이썬다운 코딩 스타일입니다. else는  지워주셔도 됩니다 :)

NoahShin replied:
네!, 저번에도 비슷한 리뷰 받았는데 또 똑같은 실수했네요 :{

soheon-lee commented on favor/views.py:
```diff
@@ -1,3 +1,65 @@
+        user_id = request.user.id
+        favor_products = [ product for product in ProductFavor.objects.filter(user=user_id).values()]
+        favor_product  = [ products['product_id'] for products in favor_products]
+
+        product_name        = [Product.objects.get(id=product_id).name for product_id in favor_product]
+        product_thumb_image = [Product.objects.get(id=product_id).thumb_image for product_id in favor_product]
+        product_price       = [Product.objects.get(id=product_id).price for product_id in favor_product]
```
Product객체를 먼저 하나 get 해서 선언하신 후, 해당 객체의 name, 해당 객체의 thumb_image, price에 접근해 보세요 :) 그럼 같은 부분을 중복해서 적는 것을 피하실 수 있습니다

NoahShin committed:
Add: ProductFavorView, SellerFavorView 작성
- 역참조해서 쿼리수를 줄일예정

soheon-lee commented on favor/views.py:
```diff
@@ -1,3 +1,100 @@
+    @login_decorator
+    def get(self, request,):
+        user_id = request.user.id
+      #  result = request.GET.get(product_id, None)  이거 가라겟겟 공부
```
더이상 사용하지 않는 주석이라면 제거해주세요 ㅎㅎ
공부 많이 하셨나요? ㅎㅎ

soheon-lee commented on product/views.py:
```diff
@@ -1,3 +1,26 @@
+
+class SearchView(View):
+    def get(self, request):
+        result = request.GET['result']
+        results = Product.objects.filter(title__contains=result)
```
```suggestion
        keyword = request.GET['result']
        products = Product.objects.filter(title__contains=keyword)
```
보다 직관적인 변수명을 사용해주시는 것이 좋습니다

soheon-lee commented on product/views.py:
```diff
@@ -1,3 +1,26 @@
+        if not results.exists():
+            return JsonResponse({'MESSAGE':'NO_RESULT!'}, status = 400)
+            
+        result_lists = [{
+            'title'            : result.title,
+            'thumb_image'      : result.thumb_image_url,
+            'discounted_price' : int(round(float(result.price) * float(1-result.sale.sale_ratio),-2)),
+            'price'            : result.price,
+            'seller'           : result.seller.name,
+            'delivery'         : result.delivery.delivery_type,
+            'sale'             : str(int(result.sale.sale_ratio * 100)) + '%'
+            } for result in results]
```
여기도 변수명 results에서 products로 변경해주세요

NoahShin committed:
Refactor: User앱 멘토 리뷰 반영

NoahShin committed:
Refactor: 멘토 리뷰 반영

NoahShin committed:
Modify: 오타 수정 및 주석 제거

NoahShin commented:
리뷰에 댓글이 왜 안달아지지..

soheon-lee commented on user/views.py:
```diff
@@ -1,3 +1,90 @@
+            nick_name = data["nick_name"]
+            password  = data["password"]
+
+            if not validate_nick_name(nick_name):
+                return JsonResponse({'message':'INVALID_NICK_NAME'}, status = 400)
+            if not validate_password(password):
+                return JsonResponse({'message':'INVALID_PASSWORD'}, status = 400)
+            if not validate_email(email):
+                return JsonResponse({'message':'INVALID_EMAIL'}, status = 400)
+            if not validate_user_name(user_name):
+                return JsonResponse({'message':'INVALID_USER_NAME'}, status = 400)
```
`if - return` 구조 goood !

NoahShin committed:
Refactor: Data.py 삭제
